src/components/molecules/ScheduleEvent/ScheduleEvent.tsx

```diff
@@ -0,0 +1,108 @@
+import React, {
+  MouseEventHandler,
+  useCallback,
+  useEffect,
+  useMemo,
+  useState,
+} from "react";
+import classNames from "classnames";
+
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+import { faBookmark as solidBookmark } from "@fortawesome/free-solid-svg-icons";
+import { faBookmark as regularBookmark } from "@fortawesome/free-regular-svg-icons";
+
+import { PersonalizedVenueEvent } from "types/venues";
+
+import { WithVenueId } from "utils/id";
+import { isTruthy } from "utils/types";
+import { isEventLive } from "utils/event";
+
+import { ONE_HOUR_IN_MINUTES } from "utils/time";
+
+import { saveEventToProfile } from "api/profile";
+import { useUser } from "hooks/useUser";
+
+import { calcStartPosition, HOUR_WIDTH } from "../Schedule/Schedule.utils";
+
+import "./ScheduleEvent.scss";
+
+export interface ScheduleEventProps {
+  event: WithVenueId<PersonalizedVenueEvent>;
+  scheduleStartHour: number;
+  isUsers?: boolean;
+}
+
+export const ScheduleEvent: React.FC<ScheduleEventProps> = ({
+  event,
+  scheduleStartHour,
+  isUsers,
+}) => {
+  const [isBookmarked, setBookmark] = useState(event.isSaved);
+
+  useEffect(() => {
+    setBookmark(event.isSaved);
+  }, [event.isSaved]);
+
+  const { userId } = useUser();
+
+  const containerClasses = useMemo(
+    () =>
+      classNames(
+        "ScheduleEvent",
+        {
+          "ScheduleEvent--live": isEventLive(event),
+        },
+        { "ScheduleEvent--users": isTruthy(isUsers) }
+      ),
+    [event, isUsers]
+  );
+
+  const eventWidth = useMemo(
+    () => (event.duration_minutes * HOUR_WIDTH) / ONE_HOUR_IN_MINUTES,
+    [event]
+  );
+  const bookmarkEvent = useCallback(() => {
+    event.isSaved = !event.isSaved;
+
+    if (userId && event.id) {
+      saveEventToProfile({
+        venueId: event.venueId,
+        userId: userId,
+        removeMode: !event.isSaved,
+        eventId: event.id,
+      });
+    }
+  }, [userId, event]);
+
+  const onBookmark: MouseEventHandler<HTMLDivElement> = useCallback(
+    (e) => {
+      e.stopPropagation();
+      bookmarkEvent();
+    },
+    [bookmarkEvent]
+  );
```

Comment:
yeah, I also don't like using `stopPropagation()`. The reason of using it here is the existence of a handler that closes schedule dropdown on every mouse click.
So there are `stopPropagation` events on event bookmark click and on weekday click. 

Probably, would be better to change the way the dropdown closes ignoring click events within Schedule block.

Reply:
> Probably, would be better to change the way the dropdown closes ignoring click events within Schedule block.

That seems like a solid approach. How hard would that be to do?

Reply:
improved click handler that hides schedule to ignore clicks within schedule area here: https://github.com/sparkletown/sparkle/pull/1302/commits/d58d8064fc749ea46f015b5c9a1c5f960daee0dd
and forgot to remove `stopPropagation` at first, done here: https://github.com/sparkletown/sparkle/pull/1302/commits/81778da4a0078c447351d7fdabae1ee2096508e5

Reply:
Hrmm.. The `e.target.closest` feels a bit hacky.

---

Found this on google:

- https://javascript.info/bubbling-and-capturing#event-target
  - The most deeply nested element that caused the event is called a target element, accessible as `event.target`.
  - `this` (`event.currentTarget`) – is the “current” element, the one that has a currently running handler on it.

More info:

- https://developer.mozilla.org/en-US/docs/Web/API/Event/target
- https://developer.mozilla.org/en-US/docs/Web/API/Event/currentTarget

I was initially thinking we could potentially compare these in this handler instead..

But then doing some more reading, it seems like `e.stopPropagation` in the lower level handler is actually the canonical/best way to do it:

- https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_bubbling_and_capture

So I guess going back to the initial code is probably the better solution here?

Reply:
@0xdevalias, I think it's better to allow event bubbling, as it's a part of specification and we might need to know somewhere else that click event happened.

If we switch back to `stopPropagation` the NavBarSchedule will close when there was a click in the "not clickable" Schedule area (not weekday, not bookmark). It means that if user clicks on the event (expecting to see the event modal, for example), the schedule will be closed.

And when we create event modal, we'll need to use `stopPropagation` again on event click. And let's say we close modal clicking outside Event Modal area but within Schedule area, both modal and schedule will be closed if we continue to use `stopPropagation` approach. Although I guess, would be better to close just Event Modal and leave Schedule opened.

just googled an an article about `stopPropagation`: https://css-tricks.com/dangers-stopping-event-propagation/

but of course, if you still think that it would be better to come back to `stopPropagation`, I'll do that.
What do you think?

Reply:
@0xdevalias @sunny-viktoryia I guess there is no perfect way to solve this, but the need to use `stopPropagation` everywhere is error-prone. I guess we should just leave a comment explaining the logic + leave the comment in .NavBarSchedule to change the name with caution & update the class here

Reply:
added the comment: https://github.com/sparkletown/sparkle/pull/1302/commits/1bc21ee7f27af654f05153ea0a2a463d476cce1e

Reply:
hold up. writing

Reply:
>  I think it's better to allow event bubbling, as it's a part of specification and we might need to know somewhere else that click event happened.

That was my original thought, but on deeper thinking, the 'deeper' event handler is attached to a button/similar, and when clicking on that button, we expect it to be handled by the button. So IMO there's no real need for it to bubble.

---

The other points you raise there are valid/relevant though.

~As an alternative question, do we even need to have a 'close `NavBarSchedule`' click handler at the 'top' anyway?~

~What happens if we remove it from here in `src/components/molecules/NavBar/NavBar.tsx`?~

```diff
<div
  className={`schedule-dropdown-backdrop ${
    isEventScheduleVisible ? "show" : ""
  }`}
-  onClick={hideEventSchedule}
 >
  <SchedulePageModal isVisible={isEventScheduleVisible} />
</div>
```

The `schedule-dropdown-backdrop` class is defined in `src/components/molecules/NavBar/NavBar.scss` and looks super hacky: basically sets itself with a fixed `position`, top/left at `0`, width/height at `100%`. Surely we have better patterns in our codebase for handling 'clicked outside the thing so please close'. Can we just use our standard `Modal` component code here rather than trying to roll our own?

---

> just googled an an article about stopPropagation: https://css-tricks.com/dangers-stopping-event-propagation/

https://css-tricks.com/dangers-stopping-event-propagation/#what-to-do-instead mentions the `event.target.closest` pattern.

I think the main thing that irked my about it is having the `.NavBarSchedule` selector here, even though it's in `src/components/molecules/NavBar/NavBar.tsx`. But reading that article, it basically needs to be a selector for the child container we want to ignore.

We could potentially add another wrapper `div` in the `NavBar` class that shares a `const` for both the selector in the `useCallback` and the `div`'s class, to keep them explicit in the same file.

That would probably be a last resort if using our default `Modal` components doesn't work, as that seems like the best solution.

Reply:
tl;dr: Preferences in order:

1. Try using standard `Modal` patterns instead
2. Wrap it in a new `div` + use same `const` for it's class + the `useCallback` with the `event.target.closest`
3. Use the existing `event.target.closest` + add a comment.

Reply:
And then potentially find other usages of `stopPropagation` and similar in the codebase currently and add `@debt` comments to them to prevent others copying them as 'good patterns'

Reply:
worked with the 2nd scenario: https://github.com/sparkletown/sparkle/pull/1302/commits/a37241a83c560a47fab9db7ae5e374bafd3cafa3

<hr />

tried to use `Modal`, but it creates blurred backdrop container to handle "out" clicks. And there is no ability to add a customize its class names saying not to blur background just for this case. It wouldn't be a problem if that backdrop didn't blur the navigation bar.

See the screenshot (changed the background color to yellow for better understanding what I'm talking about)
![image](https://user-images.githubusercontent.com/63313732/118638709-1ed9a900-b7a5-11eb-99e3-9bfbb9fec9d7.png)

Reply:
Makes sense. I'm sure there's gotta be a way to do it, but OOS of this PR. If 2 worked, let's stick with that.

---

Instead of just having a static string as you have in that commit though, make it a `const` and use that const in both the `useCallback` and the new `div` directly

Reply:
added the const: https://github.com/sparkletown/sparkle/pull/1302/commits/aa7ef18c215abd39ef422d01b27317d954d40dca